pkg/proxy/hooks/hooks.go

```diff
@@ -0,0 +1,45 @@
+// Copyright Jetstack Ltd. See LICENSE for details.
+package hooks
+
+import (
+	"fmt"
+	"sync"
+
+	k8sErrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+type Hooks struct {
+	preShutdownHooks    map[string]ShutdownHook
```

Comment:
Using a map here means there's no guaranteed order that the shutdown hooks run in. Is this intentional? If so, it'd be good to add a comment noting so to `AddPreShutdownHook`

Reply:
:+1: 
Shutdown hooks ideally should be independent of each other. I'll make a note about the order